File: solutions/875. Koko Eating Bananas/Solution.java
import java.util.*;

class Solution {
    public int minEatingSpeed(int[] piles, int h) {
        int min = 1, max = Arrays.stream(piles).max().getAsInt();
        int ans = max;

        while(min <= max){
            int mid = min + (max - min) / 2;
            long hours = 0;
            for(int pile : piles){
                hours += (pile + mid - 1) / mid;
            }
            if(hours <= h){
                max = mid - 1;
                ans = mid;
            }
            else{
                min = mid + 1;
            }
        }
        return ans;
    }
}

File: solutions/875. Koko Eating Bananas/Solution.py
class Solution:
    def minEatingSpeed(self, piles: list[int], h: int) -> int:
        min_eating_speed, max_eating_speed = 1, max(piles)
        ans = max_eating_speed

        while min_eating_speed <= max_eating_speed:
            mid = min_eating_speed + (max_eating_speed - min_eating_speed) // 2
            hours = 0
            for pile in piles:
                hours += (pile + mid - 1) // mid
            if hours <= h:
                max_eating_speed = mid - 1
                ans = mid
            else:
                min_eating_speed = mid + 1
        return ans
